### PR TITLE
NVSHAS-9568:replace built-in fanotify with upstream library

### PR DESCRIPTION
**Pull request by jayhuang-suse**
Update fanotify package with MIT-licensed upstream repo

### CODE REVIEW

File: agent/probe/faccess_linux.go [hunk 1]
```diff
@@ -818,25 +822,34 @@ func (fa *FileAccessCtrl) processEvent(ev *fsmon.EventMetadata) bool {
...
-	if events, err := fa.fanfd.GetEvents(); err == nil {
-		for _, ev := range events {
-			if dbgError := fa.fanfd.Response(ev, fa.processEvent(ev)); dbgError != nil {
-				log.WithFields(log.Fields{"dbgError": dbgError}).Debug()
-			}
-			ev.File.Close()
+	for {
+		ev, err := fa.fanfd.GetEvent(os.Getpid())
+		if err != nil || ev == nil {
+			return
```
Comment (yasker) [inline agent/probe/faccess_linux.go:832]: Error handling?

File: share/fsmon/fanotify_linux.go [hunk 2]
```diff
@@ -560,53 +563,62 @@ func (fn *FaNotify) MonitorFileEvents() {
...
+		} else { // monitor or protect mode
+			allowRead := resp && !change
+			bReporting = !allowRead // allowed app by block_access
+		}
+
+		if bReporting || change { // report changed file
+			ifile.cb(ifile.path, mask, ifile.params, pInfo)
 		}
 	}
-	return nil
```
Comment (yasker) [inline share/fsmon/fanotify_linux.go:609]: Hmm, why are we removing this line?

**Review by yasker — CHANGES_REQUESTED** [review]

File: share/fsmon/fanotify_linux.go (hunk 2 above)
Comment (jayhuang-suse) [reply share/fsmon/fanotify_linux.go:609]: My local lint complaints that it was unreachable "return" in this function.

File: agent/probe/faccess_linux.go (hunk 1 above)
Comment (jayhuang-suse) [reply agent/probe/faccess_linux.go]: Updated.

**Review by yasker — APPROVED** [review]
LGTM from coding style perspective.

File: agent/probe/faccess_linux.go [hunk 3]
```diff
@@ -857,7 +870,10 @@ func (fa *FileAccessCtrl) monitorFilePermissionEvents() {
 		}
 
 		if (pfd[0].Revents & unix.POLLIN) != 0 {
-			fa.handleEvents()
+			if err := fa.handleEvents(); err != nil && err != unix.EINTR {
+				log.WithFields(log.Fields{"err": err}).Error("FA: handle")
+				break
+			}
```
Comment (holyspectral) [inline agent/probe/faccess_linux.go:876]: In the original logic, the error from `fa.handleEvents()` will not stop the for loop.  In this new version, when error is not EINTR, the flow will leave the for loop as well as `fa.monitorFilePermissionEvents()` go routine.  Is it expected?

**Review by holyspectral — COMMENTED** [review]
Generally looks good to me.  Just a question about error handling.

File: agent/probe/faccess_linux.go (hunk 3 above)
Comment (jayhuang-suse) [reply agent/probe/faccess_linux.go:876]: It was tested by the file monitor section and is okay if we need an error handler as Sheng requested.

**Review by holyspectral — APPROVED** [review]

**Pull request merged**